[Woo POS] Coupons: Retrieve settings from storage first, remote as fallback

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-280 

## Description
This PR moves the coupon `woocommerce_enable_coupons` setting check to local storage first, and keep the remote call as a fallback, so we avoid unnecessary remote calls when not needed.

## Testing information
1. Disable coupons from store
2. Add a breakpoint in the new function, somewhere around `storageSetting`
3. Navigate to coupons in POS, observe that you're presented with an error, and prompted to enable coupons
4. In the console, storageSetting value prop shows as "no"

```
<SiteSetting: 0x60000214f700> (entity: SiteSetting; id: 0x98a9b7822442aea6 <x-coredata://F1088342-A2AB-4498-A5DA-7B00218BEA92/SiteSetting/p7>; data: {
    label = "Enable coupons";
    settingDescription = "Enable the use of coupon codes";
    settingGroupKey = general;
    settingID = "woocommerce_enable_coupons";
    siteID = "-1";
    value = no;
})
```
5. Observe the remote call and fetching the value again as a fallback
6. Tap "Enable coupons"
7. Navigate back to coupons/exiting POS/reload the app
8. Observe the entity with storageSeting value as "yes"
```
<SiteSetting: 0x6000021470c0> (entity: SiteSetting; id: 0xac0b93ee11d8b6c1 <x-coredata://F1088342-A2AB-4498-A5DA-7B00218BEA92/SiteSetting/p7>; data: {
    label = "Enable coupons";
    settingDescription = "Enable the use of coupon codes";
    settingGroupKey = general;
    settingID = "woocommerce_enable_coupons";
    siteID = "-1";
    value = yes;
})
```
9. Assume that this setting does not exist in local storage, for this change the code to:

```
let settingID = Constants.enableCouponsSettingID
var storageSetting = storage?.viewStorage.loadSiteSetting(siteID: siteID, settingID: settingID)
storageSetting = nil
```
10. Run the app again and navigate to coupons, observe how falls back back into remote check and coupons are loaded.


<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: https://woomobilep2.wordpress.com/2024/05/06/woocommerce-mobile-quality-report-march-april/#comment-12036 -->

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Check storage settings before checking remote store

iamgabrielma committed:
make constants

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -132,8 +132,25 @@ private extension PointOfSaleCouponService {
     private func checkStoreCouponSettings() async -> Bool {
-        await withCheckedContinuation { continuation in
+        let settingID = Constants.enableCouponsSettingID
+        let storageSetting = storage?.viewStorage.loadSiteSetting(siteID: siteID, settingID: settingID)
```
Turns out, no need for `storage` to be `optional` anymore. Some leftovers from previous work.

staskus commented on Yosemite/Yosemite/PointOfSale/PointOfSaleCouponService.swift:
```diff
@@ -132,8 +132,25 @@ private extension PointOfSaleCouponService {
+        let settingID = Constants.enableCouponsSettingID
+        let storageSetting = storage?.viewStorage.loadSiteSetting(siteID: siteID, settingID: settingID)
+
+        switch storageSetting {
```
nit; 

`.some` comes from `optional` but we can switch directly on the value:

```swift
        switch storageSetting?.value {
        case Constants.enableCouponsSettingValue:
            return true
        default:
            return await checkRemoteStoreCouponSettings()
        }
```

staskus committed:
Merge branch 'trunk' into task/WOOMOB-280-retrieve-coupon-setting-from-storage-first

staskus committed:
Simplify enableCouponsSetting check

staskus committed:
Check if coupons are disabled if remote syncing fails

This allows to be confident in relying on local checks for coupons setting but double-checking if the failure is related to disabled coupons

staskus committed:
Added additional PointOfSaleCouponServiceTests to confirm local setting check

staskus reviewed (approved):
@iamgabrielma Thanks. It worked well!

I know you'll be unavailable soon so I added one more case when testing: to handle a case when remote setting goes from true to false. In this case, I just check the setting when coupons request fails, to either show disabled or loading error.

staskus committed:
Merge branch 'trunk' into task/WOOMOB-280-retrieve-coupon-setting-from-storage-first

iamgabrielma commented:
> I know you'll be unavailable soon so I added one more case when testing: to handle a case when remote setting goes from true to false. In this case, I just check the setting when coupons request fails, to either show disabled or loading error.

Thanks! Appreciated 🙇
